### PR TITLE
Implement DOM to texture

### PR DESCRIPTION
**Pull request by MortimerGoro**
This is a prototype of the WebGL DOMToTexture feature. The API should be fine as a privileged extension for now due to security concerns. The working group has been investigating the viability of unprivileged usage but the spec is not ready yet.

Demo video: https://www.youtube.com/watch?v=hpZqEM5hPao

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18592)

### CODE REVIEW

**Comment by MortimerGoro** [comment]
Related PR in WR: https://github.com/servo/webrender/pull/1687

File: components/script/dom/webglrenderingcontext.rs [hunk 1]
```diff
@@ -3171,6 +3175,64 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
...
+        // Get bound texture
+        let texture = match self.bound_texture_2d.get() {
+            Some(texture) => texture,
+            None => {
+                return Ok(self.webgl_error(InvalidOperation));
+            }
+        };
+
+        // Get PipelineId from the source
+        let pipeline_id = match source.pipeline_id() {
```
Comment (jdm) [inline components/script/dom/webglrenderingcontext.rs]: `let pipeline_id = source.pipeline_id().or_or(Error::InvalidState)?;`

File: components/script/dom/webglrenderingcontext.rs [hunk 2]
```diff
@@ -3171,6 +3175,64 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
...
+        script_to_constellation_chan.send(ScriptMsg::GetDocumentId(pipeline_id, sender)).unwrap();
+
+        let document_id = match receiver.recv().unwrap() {
+            Some(document_id) => { document_id },
+            _ => {
+                return Err(Error::InvalidState);
+            }
+        };
+
+        // mark that the texture is attached to DOM.
```
Comment (jdm) [inline components/script/dom/webglrenderingcontext.rs]: I'm not sure that this comment adds anything to the code.

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by MortimerGoro** [comment]
Done!

File: components/script_traits/script_msg.rs [hunk 3]
```diff
@@ -27,7 +27,7 @@ use servo_url::ServoUrl;
 use style_traits::CSSPixel;
 use style_traits::cursor::Cursor;
 use style_traits::viewport::ViewportConstraints;
-use webrender_api::ClipId;
+use webrender_api::{ClipId, DocumentId};
```
Comment (jdm) [inline components/script_traits/script_msg.rs]: We can remove this now.

File: components/script/dom/webgltexture.rs [hunk 4]
```diff
@@ -45,6 +46,8 @@ pub struct WebGLTexture {
     mag_filter: Cell<Option<u32>>,
     #[ignore_heap_size_of = "Defined in ipc-channel"]
     renderer: WebGLMsgSender,
+    // True if this texture is used for the DOMToTexture feature.
```
Comment (jdm) [inline components/script/dom/webgltexture.rs]: nit: ///

File: components/script/dom/webglrenderingcontext.rs [hunk 5]
```diff
@@ -3171,6 +3173,45 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
...
+        // Get bound texture
+        let texture = match self.bound_texture_2d.get() {
+            Some(texture) => texture,
+            None => {
+                return Ok(self.webgl_error(InvalidOperation));
+            }
+        };
+
+        let pipeline_id = source.pipeline_id().ok_or(Error::InvalidState)?;
+        let document_id  = self.global().as_window().webrender_document();
```
Comment (jdm) [inline components/script/dom/webglrenderingcontext.rs]: We should return an error if this is invoked from a context which does not have a Window global.

File: components/script/dom/webidls/WebGLRenderingContext.webidl [hunk 6]
```diff
@@ -654,6 +654,9 @@ interface WebGLRenderingContextBase
     [Throws]
     void texImage2D(GLenum target, GLint level, GLenum internalformat,
                     GLenum format, GLenum type, TexImageSource? source); // May throw DOMException
+    [Throws, Pref="dom.webgl.dom_to_texture.enabled"]
```
Comment (jdm) [inline components/script/dom/webidls/WebGLRenderingContext.webidl]: This does not actually throw any exceptions, so we can remove the annotation.

File: components/canvas_traits/webgl.rs [hunk 7]
```diff
@@ -86,6 +88,11 @@ impl WebGLMsgSender {
         }
     }
 
+    // Returns the WebGLContextId associated to this sender
+    pub fn ctx_id(&self) -> WebGLContextId {
```
Comment (jdm) [inline components/canvas_traits/webgl.rs]: Let's call this `context_id`.

File: components/canvas_traits/webgl.rs [hunk 8]
```diff
@@ -86,6 +88,11 @@ impl WebGLMsgSender {
         }
     }
 
+    // Returns the WebGLContextId associated to this sender
```
Comment (jdm) [inline components/canvas_traits/webgl.rs]: nit: ///

**Review by jdm — CHANGES_REQUESTED** [review]
Remaining comments are all minor.

**Comment by MortimerGoro** [comment]
Done. Sorry for the delay, I was on a trip ;)

File: components/script/dom/webidls/WebGLRenderingContext.webidl (hunk 6 above)
Comment (MortimerGoro) [reply components/script/dom/webidls/WebGLRenderingContext.webidl]: It still throws the ```let pipeline_id = source.pipeline_id().or_or(Error::InvalidState)?;```exception

**Review by jdm — APPROVED** [review]

**Comment by MortimerGoro** [comment]
Merge conflict fixed

**Pull request merged**